File: ui/src/services/expressions/coordinateConversion.ts
/**
 * Coordinate Conversion Expressions
 *
 * Functions for converting between layer space, composition space, and world space.
 * Also includes lookAt and orientToPath for 3D orientation.
 */

import type { ExpressionContext } from './types';

// ============================================================
// COORDINATE CONVERSION TYPES
// ============================================================

/**
 * Transform matrix interface for coordinate conversion
 */
export interface LayerTransform {
  position: number[];
  scale: number[];
  rotation: number[];
  anchor: number[];
  parent?: LayerTransform | null;
}

// ============================================================
// 3D ORIENTATION FUNCTIONS
// ============================================================

/**
 * lookAt - Calculate rotation to face a target point
 *
 * Returns [xRotation, yRotation, zRotation] in degrees
 *
 * @param fromPoint - Source position [x, y, z]
 * @param toPoint - Target position [x, y, z]
 * @returns Rotation angles [rx, ry, rz] in degrees
 */
export function lookAt(fromPoint: number[], toPoint: number[]): number[] {
  const dx = (toPoint[0] || 0) - (fromPoint[0] || 0);
  const dy = (toPoint[1] || 0) - (fromPoint[1] || 0);
  const dz = (toPoint[2] || 0) - (fromPoint[2] || 0);

  // Calculate yaw (Y rotation) and pitch (X rotation)
  const yaw = Math.atan2(dx, dz) * 180 / Math.PI;
  const dist = Math.sqrt(dx * dx + dz * dz);
  const pitch = -Math.atan2(dy, dist) * 180 / Math.PI;

  return [pitch, yaw, 0];
}

/**
 * orientToPath - Auto-orient layer along motion path
 *
 * Given position keyframes, calculates rotation to face movement direction
 */
export function orientToPath(
  ctx: ExpressionContext,
  tangentVector?: number[]
): number[] {
  // If tangent provided, use it directly
  if (tangentVector) {
    const [dx, dy, dz] = tangentVector;
    const yaw = Math.atan2(dx || 0, dz || 1) * 180 / Math.PI;
    const dist = Math.sqrt((dx || 0) ** 2 + (dz || 1) ** 2);
    const pitch = -Math.atan2(dy || 0, dist) * 180 / Math.PI;
    return [pitch, yaw, 0];
  }

  // Calculate tangent from velocity
  const vel = ctx.velocity;
  if (Array.isArray(vel) && vel.length >= 2) {
    const dx = vel[0] || 0;
    const dy = vel[1] || 0;
    const dz = vel[2] || 0;
    const yaw = Math.atan2(dx, dz || 1) * 180 / Math.PI;
    const dist = Math.sqrt(dx ** 2 + (dz || 1) ** 2);
    const pitch = -Math.atan2(dy, dist) * 180 / Math.PI;
    return [pitch, yaw, 0];
  }

  return [0, 0, 0];
}

// ============================================================
// COORDINATE CONVERSION FUNCTIONS
// ============================================================

// Maximum recursion depth for parent traversal (BUG-010: prevent stack overflow)
const MAX_PARENT_DEPTH = 50;

/**
 * toComp - Convert point from layer space to composition space
 *
 * @param point - Point in layer coordinates [x, y] or [x, y, z]
 * @param layerTransform - The layer's transform properties
 * @param depth - Current recursion depth (for cycle protection)
 * @returns Point in composition coordinates
 */
export function toComp(point: number[], layerTransform: LayerTransform, depth: number = 0): number[] {
  // Guard against circular parent chains (BUG-010)
  if (depth > MAX_PARENT_DEPTH) {
    console.warn('[Expressions] Max parent depth exceeded in toComp - possible circular reference');
    return point;
  }

  const [px, py, pz = 0] = point;
  const { position, scale, rotation, anchor } = layerTransform;

  // Apply anchor offset
  let x = px - (anchor[0] || 0);
  let y = py - (anchor[1] || 0);
  let z = pz - (anchor[2] || 0);

  // Apply scale (BUG-011: use ?? to preserve intentional 0)
  x *= (scale[0] ?? 100) / 100;
  y *= (scale[1] ?? 100) / 100;
  z *= (scale[2] ?? 100) / 100;

  // Apply rotation (Z, then Y, then X - matching AE order)
  const rz = (rotation[2] || rotation[0] || 0) * Math.PI / 180;
  const ry = (rotation[1] || 0) * Math.PI / 180;
  const rx = (rotation[0] || 0) * Math.PI / 180;

  // Rotate around Z
  let x1 = x * Math.cos(rz) - y * Math.sin(rz);
  let y1 = x * Math.sin(rz) + y * Math.cos(rz);
  let z1 = z;

  // Rotate around Y (3D only)
  let x2 = x1 * Math.cos(ry) + z1 * Math.sin(ry);
  let y2 = y1;
  let z2 = -x1 * Math.sin(ry) + z1 * Math.cos(ry);

  // Rotate around X (3D only)
  let x3 = x2;
  let y3 = y2 * Math.cos(rx) - z2 * Math.sin(rx);
  let z3 = y2 * Math.sin(rx) + z2 * Math.cos(rx);

  // Apply position offset
  x3 += position[0] || 0;
  y3 += position[1] || 0;
  z3 += position[2] || 0;

  // Recursively apply parent transforms
  if (layerTransform.parent) {
    return toComp([x3, y3, z3], layerTransform.parent, depth + 1);
  }

  return point.length === 2 ? [x3, y3] : [x3, y3, z3];
}

/**
 * fromComp - Convert point from composition space to layer space
 *
 * @param point - Point in composition coordinates [x, y] or [x, y, z]
 * @param layerTransform - The layer's transform properties
 * @param depth - Current recursion depth (for cycle protection)
 * @returns Point in layer coordinates
 */
export function fromComp(point: number[], layerTransform: LayerTransform, depth: number = 0): number[] {
  // Guard against circular parent chains (BUG-010)
  if (depth > MAX_PARENT_DEPTH) {
    console.warn('[Expressions] Max parent depth exceeded in fromComp - possible circular reference');
    return point;
  }

  let [px, py, pz = 0] = point;

  // First, if there's a parent, convert from comp to parent space
  if (layerTransform.parent) {
    [px, py, pz] = fromComp([px, py, pz], layerTransform.parent, depth + 1);
  }

  const { position, scale, rotation, anchor } = layerTransform;

  // Subtract position
  let x = px - (position[0] || 0);
  let y = py - (position[1] || 0);
  let z = pz - (position[2] || 0);

  // Inverse rotation (X, then Y, then Z - reverse order)
  const rz = -(rotation[2] || rotation[0] || 0) * Math.PI / 180;
  const ry = -(rotation[1] || 0) * Math.PI / 180;
  const rx = -(rotation[0] || 0) * Math.PI / 180;

  // Rotate around X (inverse)
  let x1 = x;
  let y1 = y * Math.cos(rx) - z * Math.sin(rx);
  let z1 = y * Math.sin(rx) + z * Math.cos(rx);

  // Rotate around Y (inverse)
  let x2 = x1 * Math.cos(ry) + z1 * Math.sin(ry);
  let y2 = y1;
  let z2 = -x1 * Math.sin(ry) + z1 * Math.cos(ry);

  // Rotate around Z (inverse)
  let x3 = x2 * Math.cos(rz) - y2 * Math.sin(rz);
  let y3 = x2 * Math.sin(rz) + y2 * Math.cos(rz);
  let z3 = z2;

  // Inverse scale (BUG-011: use ?? to preserve intentional 0, || 1 guards div/0)
  const sx = (scale[0] ?? 100) / 100;
  const sy = (scale[1] ?? 100) / 100;
  const sz = (scale[2] ?? 100) / 100;
  if (sx === 0 || sy === 0 || sz === 0) {
    console.warn('[Expressions] Scale of 0 in fromComp produces undefined inverse');
  }
  x3 /= sx || 1;
  y3 /= sy || 1;
  z3 /= sz || 1;

  // Add anchor
  x3 += anchor[0] || 0;
  y3 += anchor[1] || 0;
  z3 += anchor[2] || 0;

  return point.length === 2 ? [x3, y3] : [x3, y3, z3];
}

/**
 * toWorld - Convert point from layer space to world (3D) space
 * Alias for toComp in 3D context
 */
export function toWorld(point: number[], layerTransform: LayerTransform): number[] {
  // Ensure 3D point
  const point3D = point.length === 2 ? [...point, 0] : point;
  return toComp(point3D, layerTransform);
}

/**
 * fromWorld - Convert point from world space to layer space
 * Alias for fromComp in 3D context
 */
export function fromWorld(point: number[], layerTransform: LayerTransform): number[] {
  // Ensure 3D point
  const point3D = point.length === 2 ? [...point, 0] : point;
  return fromComp(point3D, layerTransform);
}

// ============================================================
// COORDINATE CONVERSION NAMESPACE
// ============================================================

/**
 * Coordinate conversion namespace
 */
export const coords = {
  toComp,
  fromComp,
  toWorld,
  fromWorld,
  lookAt,
  orientToPath,
};

File: ComfyUI-Lattice-Compositor/leancomfy/haskell/Lattice/Services/CoordinateConversion.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}

{-|
Module      : Lattice.Services.CoordinateConversion
Description : Coordinate conversion math
Copyright   : (c) Lattice, 2026
License     : MIT

Functions for converting between coordinate spaces and calculating
orientations. Pure math without side effects.

Source: ui/src/services/expressions/coordinateConversion.ts
-}

module Lattice.Services.CoordinateConversion
  ( -- * Types
    Point3D(..)
  , EulerRotation(..)
  , LayerTransform(..)
    -- * Constants
  , degToRad, radToDeg
    -- * Orientation
  , lookAt, orientToTangent
    -- * Rotation Helpers
  , rotateX, rotateY, rotateZ
    -- * Coordinate Conversion
  , toCompSingle, fromCompSingle
  , toComp, fromComp
  , toWorld, fromWorld
    -- * Utilities
  , safeDivide
  , maxParentDepth
  ) where

import GHC.Generics (Generic)

--------------------------------------------------------------------------------
-- Constants
--------------------------------------------------------------------------------

-- | Convert degrees to radians
degToRad :: Double -> Double
degToRad deg = deg * pi / 180

-- | Convert radians to degrees
radToDeg :: Double -> Double
radToDeg rad = rad * 180 / pi

--------------------------------------------------------------------------------
-- Types
--------------------------------------------------------------------------------

-- | 3D point
data Point3D = Point3D
  { point3DX :: Double
  , point3DY :: Double
  , point3DZ :: Double
  } deriving stock (Eq, Show, Generic)

-- | Euler rotation angles in degrees
data EulerRotation = EulerRotation
  { eulerPitch :: Double  -- ^ X rotation
  , eulerYaw :: Double    -- ^ Y rotation
  , eulerRoll :: Double   -- ^ Z rotation
  } deriving stock (Eq, Show, Generic)

-- | Layer transform properties
data LayerTransform = LayerTransform
  { ltPosition :: Point3D
  , ltScale :: Point3D     -- ^ As percentages (100 = 1.0)
  , ltRotation :: Point3D  -- ^ Euler angles in degrees
  , ltAnchor :: Point3D
  } deriving stock (Eq, Show, Generic)

-- | Identity transform
identityTransform :: LayerTransform
identityTransform = LayerTransform
  { ltPosition = Point3D 0 0 0
  , ltScale = Point3D 100 100 100
  , ltRotation = Point3D 0 0 0
  , ltAnchor = Point3D 0 0 0
  }

--------------------------------------------------------------------------------
-- Orientation Functions
--------------------------------------------------------------------------------

-- | Calculate rotation to face from one point to another.
--   Returns rotation angles (pitch, yaw, roll) in degrees.
--   Uses Y-up, Z-forward convention.
lookAt :: Point3D -> Point3D -> EulerRotation
lookAt from_ to =
  let dx = point3DX to - point3DX from_
      dy = point3DY to - point3DY from_
      dz = point3DZ to - point3DZ from_
      -- Calculate yaw (Y rotation) from XZ plane
      yaw = radToDeg (atan2 dx dz)
      -- Calculate pitch (X rotation) from horizontal distance
      dist = sqrt (dx * dx + dz * dz)
      pitch = radToDeg (-(atan2 dy dist))
  in EulerRotation pitch yaw 0

-- | Calculate rotation from velocity/tangent vector
orientToTangent :: Point3D -> EulerRotation
orientToTangent tangent =
  let dx = if isNaN (point3DX tangent) || isInfinite (point3DX tangent)
           then 0 else point3DX tangent
      dy = if isNaN (point3DY tangent) || isInfinite (point3DY tangent)
           then 0 else point3DY tangent
      dz = if isNaN (point3DZ tangent) || isInfinite (point3DZ tangent)
           then 1 else point3DZ tangent  -- Default forward
      yaw = radToDeg (atan2 dx dz)
      dist = sqrt (dx * dx + dz * dz)
      pitch = radToDeg (-(atan2 dy dist))
  in EulerRotation pitch yaw 0

--------------------------------------------------------------------------------
-- Rotation Helpers
--------------------------------------------------------------------------------

-- | Apply rotation around Z axis
rotateZ :: Double -> Double -> Double -> Double -> Point3D
rotateZ x y z angle =
  let rad = degToRad angle
      cosA = cos rad
      sinA = sin rad
  in Point3D (x * cosA - y * sinA) (x * sinA + y * cosA) z

-- | Apply rotation around Y axis
rotateY :: Point3D -> Double -> Point3D
rotateY p angle =
  let rad = degToRad angle
      cosA = cos rad
      sinA = sin rad
  in Point3D
       (point3DX p * cosA + point3DZ p * sinA)
       (point3DY p)
       (-point3DX p * sinA + point3DZ p * cosA)

-- | Apply rotation around X axis
rotateX :: Point3D -> Double -> Point3D
rotateX p angle =
  let rad = degToRad angle
      cosA = cos rad
      sinA = sin rad
  in Point3D
       (point3DX p)
       (point3DY p * cosA - point3DZ p * sinA)
       (point3DY p * sinA + point3DZ p * cosA)

--------------------------------------------------------------------------------
-- Coordinate Conversion (Single Layer)
--------------------------------------------------------------------------------

-- | Safe divide with fallback
safeDivide :: Double -> Double -> Double -> Double
safeDivide num denom fallback
  | denom == 0 || isNaN denom || isInfinite denom = fallback
  | otherwise =
      let result = num / denom
      in if isNaN result || isInfinite result then fallback else result

-- | Convert point from layer space to composition space (single layer).
--   Applies: anchor offset → scale → rotation (ZYX order) → position
toCompSingle :: Point3D -> LayerTransform -> Point3D
toCompSingle point transform =
  -- Apply anchor offset
  let x = point3DX point - point3DX (ltAnchor transform)
      y = point3DY point - point3DY (ltAnchor transform)
      z = point3DZ point - point3DZ (ltAnchor transform)
      -- Apply scale (percentage to multiplier)
      x' = x * point3DX (ltScale transform) / 100
      y' = y * point3DY (ltScale transform) / 100
      z' = z * point3DZ (ltScale transform) / 100
      -- Apply rotation (Z, then Y, then X - matching AE order)
      rotZ = rotateZ x' y' z' (point3DZ (ltRotation transform))
      rotY = rotateY rotZ (point3DY (ltRotation transform))
      rotX = rotateX rotY (point3DX (ltRotation transform))
  -- Apply position offset
  in Point3D
       (point3DX rotX + point3DX (ltPosition transform))
       (point3DY rotX + point3DY (ltPosition transform))
       (point3DZ rotX + point3DZ (ltPosition transform))

-- | Convert point from composition space to layer space (single layer).
--   Applies: position → inverse rotation (XYZ order) → inverse scale → anchor
fromCompSingle :: Point3D -> LayerTransform -> Point3D
fromCompSingle point transform =
  -- Subtract position
  let x = point3DX point - point3DX (ltPosition transform)
      y = point3DY point - point3DY (ltPosition transform)
      z = point3DZ point - point3DZ (ltPosition transform)
      -- Inverse rotation (X, then Y, then Z - reverse order, negative angles)
      rotX = rotateX (Point3D x y z) (-(point3DX (ltRotation transform)))
      rotY = rotateY rotX (-(point3DY (ltRotation transform)))
      rotZ = rotateZ (point3DX rotY) (point3DY rotY) (point3DZ rotY)
                     (-(point3DZ (ltRotation transform)))
      -- Inverse scale (divide, with safety)
      sx = point3DX (ltScale transform) / 100
      sy = point3DY (ltScale transform) / 100
      sz = point3DZ (ltScale transform) / 100
      x' = safeDivide (point3DX rotZ) sx (point3DX rotZ)
      y' = safeDivide (point3DY rotZ) sy (point3DY rotZ)
      z' = safeDivide (point3DZ rotZ) sz (point3DZ rotZ)
  -- Add anchor
  in Point3D
       (x' + point3DX (ltAnchor transform))
       (y' + point3DY (ltAnchor transform))
       (z' + point3DZ (ltAnchor transform))

--------------------------------------------------------------------------------
-- Coordinate Conversion with Parent Chain
--------------------------------------------------------------------------------

-- | Maximum parent depth to prevent infinite loops
maxParentDepth :: Int
maxParentDepth = 50

-- | Convert to composition space with parent chain
toComp :: Point3D -> [LayerTransform] -> Point3D
toComp point transforms = go point transforms maxParentDepth
  where
    go p [] _ = p
    go p _ 0 = p  -- Max depth reached
    go p (t:rest) n =
      let transformed = toCompSingle p t
      in go transformed rest (n - 1)

-- | Convert from composition space with parent chain
fromComp :: Point3D -> [LayerTransform] -> Point3D
fromComp point transforms = go point (reverse transforms) maxParentDepth
  where
    go p [] _ = p
    go p _ 0 = p  -- Max depth reached
    go p (t:rest) n =
      let parentTransformed = go p rest n
      in fromCompSingle parentTransformed t

--------------------------------------------------------------------------------
-- Aliases for 3D Context
--------------------------------------------------------------------------------

-- | Convert to world space (alias for toComp)
toWorld :: Point3D -> [LayerTransform] -> Point3D
toWorld = toComp

-- | Convert from world space (alias for fromComp)
fromWorld :: Point3D -> [LayerTransform] -> Point3D
fromWorld = fromComp
